ABM3 Intra-Household AV Allocation and Taxi/TNC Routing Models

dhensle opened the pull request:
**Overview**
This PR implements both the intra-household AV allocation and taxi/TNC routing models for the ABM3 system.

**Implementation Details**
- Intra-household AV allocation model
- Taxi/TNC routing algorithm with optimized vehicle assignment
- Configurable parameters for vehicle occupancy, wait times, and batch processing
- Performance optimizations for large households/trips

**Testing Results & Model Performance**
The TNC routing model has been tested and validated with the following performance metrics:

TNC Trip Statistics
- Total TNC Vehicle Trips: 413,714
- Total TNC Trips: 205,277
- Total Vehicles: 7,648
- TNC Trips to Vehicles Ratio: 26.84
- Average Trips per Vehicle: 54.09

Service Quality Metrics
- Average Initial Wait Time: 2.60 minutes
- Average Detour Time per Passenger (pooled trips): 4.28 minutes

Vehicle Efficiency Metrics
- Percentage of Deadhead Vehicle Trips: 50.31%
- Average Occupancy (all vehicle trips): 0.95
- Average Occupancy (weighted by VMT): 1.62
- Total Refuel Trips: 3,160 (0.76% of all vehicle trips)

VMT Analysis
- Total VMT from TNC Vehicle Trips: 1,708,092.80 miles
- Total Deadhead VMT: 235,714.59 miles
- Percentage of Deadhead VMT: 13.80%

For runtime improvements, here is a quick comparison from Iteration 3 (100% sample). In total, the 
- Java AV allocation model + TNC routing model: 0:36:45
- Python Taxi/TNC routing model + AV and TNC matrix builder: 0:02:30

This translates to a total runtime savings of roughly 90–100 minutes for the AV/TNC routing components across all three iterations.

Below shows the summary comparing between new Python algorithm and old Java algorithm:
<img width="294" height="68" alt="image" src="https://github.com/user-attachments/assets/5bbe2217-09a9-4c0c-8f3b-2de731d580cf" />
The **new Python algorithm** shows significantly improved fleet efficiency but with trade-offs in service quality:
Key Improvements ✓
- 57% reduction in total vehicles needed (7,648 vs 17,635)
- 147% increase in average trips per vehicle (54.09 vs 21.89)
- 13% increase in trips-to-vehicles ratio (26.84 vs 23.76)

Performance Trade-offs ⚠️
- 27% increase in deadhead trips (50.31% vs 39.45%)
- 13% decrease in average occupancy (0.95 vs 1.09)
- 7% increase in total vehicle trips (413,714 vs 386,084)

The new python algorithm does not have significant impact to mode share and regional VMT, as shown in the figure below from Model Comparison Report:
<img width="905" height="108" alt="image" src="https://github.com/user-attachments/assets/fa5e534c-45d8-4ad2-af85-4e3c81751dc0" /> 

<img width="1115" height="626" alt="image" src="https://github.com/user-attachments/assets/d9619f11-72e5-4fe6-8d68-da18e53d8b33" />


The new Python algorithm also fixes the bug in Java program (below was from RSG developing work):
<img width="1480" height="733" alt="image" src="https://github.com/user-attachments/assets/6b86f778-8e15-461a-baf3-49e896393776" />

dhensle committed:
initial av_trip_matching code

dhensle committed:
Merge branch 'main' of https://github.com/SANDAG/ABM into ABM3_AV_TNC_routing

dhensle committed:
blacken and correct availability conditions

dhensle committed:
av_repositioning with next trip alternatives

dhensle committed:
Merge branch 'main' of https://github.com/SANDAG/ABM into ABM3_AV_TNC_routing

dhensle committed:
running with av_repositioning

dhensle committed:
missed import in extensions init

aletzdy committed:
updated utils for av repositioning

dhensle committed:
initial taxi tnc routing commit

dhensle committed:
add timing to tnc routing

dhensle committed:
Merge branch 'ABM3_AV_TNC_routing' of https://github.com/SANDAG/ABM into ABM3_AV_TNC_routing

dhensle committed:
working av_repositioning configs

dhensle committed:
vehicle to trip matching and creation of veh trips

dhensle committed:
runtime optimization, multiple trips on tour in time period, trip matching repositioning

dhensle committed:
settings, test scenario code

dhensle committed:
tnc vehicle refueling

dhensle committed:
skim data periods, maz in outputs

dhensle committed:
updating av_repositioning to handle multiple trips in tour in time period

dhensle committed:
fixing av_repositioning spec bug

dhensle committed:
fixing refuel bug

dhensle committed:
fixing deadheading labeling and repositioning choosers

dhensle committed:
repositioning spec updates

JiaXu1024 reviewed (commented):
The AV allocation model is not directly integrated with the main ActivitySim model pipeline. Could you update the config file to include it in the workflow?

dhensle committed:
summary plots, special market demand

dhensle committed:
adding distance, batching, output id map, and bug fixes

dhensle committed:
taxi tnc settings yaml update

dhensle committed:
formatting

dhensle committed:
adding vehicle occupancy

dhensle committed:
removing unnecessary dummy coefficients

dhensle committed:
Adding docstrings

dhensle committed:
fixing hard-coded simulation time bin in refueling

dhensle committed:
deleting unused argument from check_refuel_needs

dhensle committed:
fixing bug looking at time instead of dist for refuel

JiaXu1024 commented on src/asim/scripts/taxi_tnc_routing/TNC_summaries_and_comparison.ipynb:
```diff
@@ -0,0 +1,15750 @@
+    "\n",
+    "landuse_file = r\"C:\\projects\\sandag\\av_tnc_routing\\av_run_dir\\tnc_routing_data\\final_land_use.csv\"\n",
+    "omx_path = r\"C:\\projects\\sandag\\av_tnc_routing\\av_run_dir\\input_data_full\\skims\\traffic_skims_MD.omx\"\n",
+    "skim_core = \"SOV_TR_H_DIST__MD\""
```
The skim_core is different from settings as default setting is shared-ride 2 high value-of-time skim.

JiaXu1024 commented on src/asim/scripts/taxi_tnc_routing/taxi_tnc_routing.py:
```diff
@@ -0,0 +1,1800 @@
+    def write_outputs(self, tnc_veh_trips, pooled_trips, id_mapping_df):
+        """Write final tables to output folder"""
+        tnc_veh_trips.to_csv(
+            os.path.join(self.settings.output_dir, "output_tnc_vehicle_trips.csv"),
```
Is this output file equivalent to the old TNCTrips.csv? If yes, could you rename it to use the same file name as before?

JiaXu1024 commented:
Were the demand matrix files produced using the updated model as well? i.e. TNCVehicleTrips_[time_period].omx and emptyAVTrips.omx

dhensle committed:
implement chunking in av_trip_matching

dhensle committed:
adding explicit_chunk setting

dhensle committed:
initial commit of tnc_av_matrix_builder.py

dhensle committed:
improve mem by dropping unused cols in av_routing

dhensle committed:
blacken taxi_tnc_routing.py

dhensle committed:
adding documentation to ABM3 wiki

dhensle committed:
Adding optional settings for max trips and avs to consider

dhensle committed:
add user setting to specify the max number of next trips to consider for repositioning

dhensle commented on src/asim/extensions/av_routing.py:
```diff
@@ -0,0 +1,1079 @@
+    return skims, interaction_df
+
+
+def construct_av_to_trip_alternatives(num_avs, num_trips):
```
Yes, good idea. Added!

dhensle committed:
fixing pandas future warnings on fillna

dhensle committed:
updating default settings for taxi_tnc_routing

dhensle committed:
adding av_routing to settings yamls

dhensle committed:
excluding av trips from parking location choice model

dhensle committed:
updating path settings to match ABM3 expectations

dhensle committed:
sensitivity test summary notebooks

dhensle committed:
Merge remote-tracking branch 'origin/main' into ABM3_AV_TNC_routing

dhensle committed:
integrating updated MAAS into master run

dhensle commented on src/asim/configs/resident/av_repositioning.csv:
```diff
@@ -0,0 +1,31 @@
+# service next trip util 1,,,,,,,,
+util_next_trip_available,Unavailable if no next trip,next_trip_id_1 == -1,,,,coef_unavailable,,
+util_next1_ivt,in vehicle time to next trip 1 origin,v_to_trip_orig1_time,,,,coef_ivt,,
+util_next1_rel,reliability vehicle location to next trip 1 origin,"@df.v_to_trip_orig1_rel * 14 / np.maximum(0.1, df.v_to_trip_orig1_dist)",,,,coef_ivt,,
```
Keeping this consistent with the Java uec.

dhensle commented on docs/design/demand/resident.md:
```diff
@@ -88,5 +88,5 @@ flowchart TD
 
 
-At this point, all tours are generated, scheduled, have a primary destination, and a selected tour mode. The next set of models fills in details about the tours - number of intermediate stops, location of each stop, the departure time of each stop, and the mode of each trip on the tour. Finally, the parking location of each auto trip to the central business district (CBD) is determined.
+At this point, all tours are generated, scheduled, have a primary destination, and a selected tour mode. The next set of models fills in details about the tours - number of intermediate stops, location of each stop, the departure time of each stop, and the mode of each trip on the tour. For households that own autonomous vehicles, an additional [AV Routing Model](av-routing.md) is run after the r ...
```
This is actually not a run-on sentence.  Ignoring suggestion.

dhensle commented:
@JiaXu1024 I believe I have addressed all of the comments here.  We are doing a final integration test and will of course fix any issues that arise from that, but I believe this is functionally complete!

dhensle committed:
Add sample rate argument to matrix builder

dhensle committed:
putting all tnc routing settings in the yaml file

dhensle committed:
updating checks on written vs inputs with sample rate

JiaXu1024 committed:
Merge branch 'main' into ABM3_AV_TNC_routing

bhargavasana commented on src/asim/configs/resident/av_repositioning.csv:
```diff
@@ -0,0 +1,31 @@
+util_time_go_to_parking,Remote- Need car soon,"@np.where(df.duration_hrs < df.AV_maxDuration, (-1)*(df.AV_maxBenefit + (df.duration_hrs*df.slope)), 0)",,coef_ivt,,,,
+util_remote_cost,Remote park - Cost of parking until departure,duration_hrs * RemoteParkingCostPerHour * 100,,coef_cost,,,,
+util_remote_ivt,in vehicle time to remote parking location,@v_to_parking_skim['SOV_TR_H_TIME'],,coef_ivt,,,,
+util_remote_rel,reliability vehicle location to remote parking location,"@v_to_parking_skim['SOV_TR_H_REL'] * 14 / np.maximum(0.1, v_to_parking_skim['SOV_TR_H_DIST'])",,coef_ivt,,,,
```
What exactly is the factor 14 we are multiplying reliability by?

dhensle replied:
This is the same reliability factor we use in tour mode choice. Added comment to spec file.

dhensle committed:
align settings

dhensle committed:
addressing review comments

dhensle commented on docs/design/demand/resident.md:
```diff
@@ -88,5 +88,5 @@ flowchart TD
 
 
-At this point, all tours are generated, scheduled, have a primary destination, and a selected tour mode. The next set of models fills in details about the tours - number of intermediate stops, location of each stop, the departure time of each stop, and the mode of each trip on the tour. Finally, the parking location of each auto trip to the central business district (CBD) is determined.
+At this point, all tours are generated, scheduled, have a primary destination, and a selected tour mode. The next set of models fills in details about the tours - number of intermediate stops, location of each stop, the departure time of each stop, and the mode of each trip on the tour. For households that own autonomous vehicles, an additional [AV Routing Model](av-routing.md) is run after the r ...
```
Yes, corrected

dhensle commented on src/asim/configs/resident/av_repositioning.csv:
```diff
@@ -0,0 +1,31 @@
+util_next1_rel,reliability vehicle location to next trip 1 origin,"@df.v_to_trip_orig1_rel * 14 / np.maximum(0.1, df.v_to_trip_orig1_dist)",,,,coef_ivt,,
+util_next1_auto_cost,auto operating cost from vehicle location to next trip 1 origin,@(df.v_to_trip_orig1_dist * costPerMile * autoCostPerMileFactorAV) + df.v_to_trip_orig1_toll,,,,coef_cost,,
+# service next trip util 2,,,,,,,,
+util_next_trip_available,Unavailable if no next trip,next_trip_id_2 == -1,,,,,coef_unavailable,
```
yes, fixed

dhensle commented on src/asim/configs/resident/av_repositioning_preprocessor.csv:
```diff
@@ -0,0 +1,50 @@
+Description,Target,Expression
+next trip start time,next_depart,df['trip_id'].map(trips.groupby('tour_id')['depart'].shift(-1).to_dict())
+trip duration,duration_hrs,"np.where(next_depart.isna(), 0, next_depart - df.depart) / 2"
+duration benefit,duration_benefit,"np.where((duration_hrs < 1.5) & (duration_hrs > 0), 60 - (60/1.5 * duration_hrs), 0)"
```
dropped

dhensle commented on src/asim/configs/resident/av_repositioning_preprocessor.csv:
```diff
@@ -0,0 +1,50 @@
+Effective parking cost,parkingCost,"np.where(is_joint+is_indiv*(1-reimbursePurpose), parkingCostBeforeReimb,_parkingCost)"
+Parking cost is 0 if going home,parkingCost,"np.where(df.destination == df.home_zone_id, 0, parkingCost)"
+# These following two are from AutonomousVehicleAllocationChoice.xls,,
+Maximim benefit for keeping car close (min),AV_maxBenefit,60
```
Added these to constants.yaml and updated the expressions for av_trip_matching and av_repositioning

dhensle commented on src/asim/configs/resident/av_trip_matching_preprocessor.csv:
```diff
@@ -0,0 +1,4 @@
+Description,Target,Expression
+next trip start time,next_depart,df['trip_id'].map(trips.groupby('tour_id')['depart'].shift(-1).to_dict())
+trip duration,duration_hrs,"np.where(next_depart.isna(), 0, next_depart - df.depart) * 2"
```
yes, fixed

dhensle commented on src/asim/configs/resident/av_trip_matching_preprocessor.csv:
```diff
@@ -0,0 +1,4 @@
+Description,Target,Expression
+next trip start time,next_depart,df['trip_id'].map(trips.groupby('tour_id')['depart'].shift(-1).to_dict())
+trip duration,duration_hrs,"np.where(next_depart.isna(), 0, next_depart - df.depart) * 2"
+duration benefit,duration_benefit,"np.where((duration_hrs < 1.5) & (duration_hrs > 0), 60 - (60/1.5 * duration_hrs), 0)"
```
Yes, fixed.  See above comment.

bhargavasana commented on src/asim/scripts/taxi_tnc_routing/taxi_tnc_routing.py:
```diff
@@ -0,0 +1,1800 @@
+
+class TaxiTNCSettings(BaseModel):
+    """
+    Taxi / TNC route choice settings
```
'pair_batch_size' and 'max_wait_time' added in https://github.com/SANDAG/ABM/pull/366/changes/25ba925a5fd2d8de9e70a9fde969daf80066b6cf

'max_vehicle_occupancy' added in https://github.com/SANDAG/ABM/pull/366/changes/b77c6b78c09da3dbae77903b0c46d8c5c6b61687

JiaXu1024 committed:
merged main branch

JiaXu1024 committed:
update environment to asim_151 to run MAAS

JiaXu1024 committed:
remove duplicate key AV_maxBenefit and AV_maxDuration in constants config file

JiaXu1024 committed:
added iteration=msa_iteration to MAAS run_proc to generate log files for each iteration

JiaXu1024 committed:
to differentiate between resident, visitor, and crossborder trips by including the parent directory name when the basename is generic
